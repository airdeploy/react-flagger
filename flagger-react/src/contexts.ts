import {IEntity, IFlaggerConfiguration} from 'flagger/types/Types'
import React from 'react'

export interface IFlaggerCtx {
  entity?: IEntity
  config?: IFlaggerConfiguration
  loading: boolean
  getVariation: (codename: string, entity?: IEntity) => string
  getFlagDetails: (
    codename: string,
    entity?: IEntity
  ) => {variation: string; enabled: boolean; isSampled: boolean; payload: any}
}

export const flaggerCtx = React.createContext<IFlaggerCtx>({
  loading: false,
  getVariation: () => 'off',
  getFlagDetails: () => ({
    variation: 'off',
    enabled: false,
    isSampled: false,
    payload: null,
  }),
})
