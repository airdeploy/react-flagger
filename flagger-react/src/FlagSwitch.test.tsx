import Flagger from 'flagger'
import React from 'react'
import renderer from 'react-test-renderer'

import {flaggerCtx} from './contexts'
import {FlagSwitch} from './index'

jest.mock('flagger')

describe('<FlagSwitch> tests', () => {
  beforeEach(async () => {
    jest.clearAllMocks()
    ;(Flagger as any).mockClear()
  })

  afterAll(async () => {
    jest.clearAllMocks()
  })

  test('accepts function as children', () => {
    const component = renderer.create(
      <FlagSwitch flag="example">
        {(props: any) => <div>{JSON.stringify(props, null, 2)}</div>}
      </FlagSwitch>
    )

    expect(component.toJSON()).toMatchSnapshot()
  })

  test('accepts react component as "component" prop', () => {
    const TestComponent = (props: any) => (
      <div>{JSON.stringify(props, null, 2)}</div>
    )
    const component = renderer.create(
      <FlagSwitch flag="example" component={TestComponent} />
    )

    expect(component.toJSON()).toMatchSnapshot()
  })

  test('gets "loading" from flaggerCtx', () => {
    const TestComponent = (props: any) => (
      <div>{JSON.stringify(props, null, 2)}</div>
    )
    const component = renderer.create(
      <flaggerCtx.Provider
        value={{
          loading: true,
          getVariation: jest.fn(() => 'off'),
          getFlagDetails: jest.fn(),
        }}>
        <FlagSwitch flag="example" component={TestComponent} />
      </flaggerCtx.Provider>
    )
    const expectedRender = renderer.create(
      <TestComponent flag="example" variation="off" loading={true} />
    )

    expect(component.toJSON()).toMatchSnapshot()
    expect(component.toJSON()).toEqual(expectedRender.toJSON())
  })

  test('calls "getVariation()" from flaggerCtx', () => {
    const getVariation = jest.fn(() => 'on')
    const TestComponent = (props: any) => (
      <div>{JSON.stringify(props, null, 2)}</div>
    )
    const component = renderer.create(
      <flaggerCtx.Provider
        value={{
          loading: false,
          getVariation,
          getFlagDetails: jest.fn(),
        }}>
        <FlagSwitch flag="example" component={TestComponent} />
      </flaggerCtx.Provider>
    )

    expect(component.toJSON()).toMatchSnapshot()
    expect(getVariation).toBeCalledTimes(1)
    expect(getVariation.mock.calls[0]).toEqual(['example', undefined])
  })

  test('defaults to entity from flaggerCtx', () => {
    const getVariation = jest.fn(() => 'on')
    const defaultEntity = {id: '42'}
    const TestComponent = (props: any) => (
      <div>{JSON.stringify(props, null, 2)}</div>
    )
    const component = renderer.create(
      <flaggerCtx.Provider
        value={{
          entity: defaultEntity,
          loading: false,
          getVariation,
          getFlagDetails: jest.fn(),
        }}>
        <FlagSwitch flag="example" component={TestComponent} />
      </flaggerCtx.Provider>
    )

    expect(component.toJSON()).toMatchSnapshot()
    expect(getVariation).toBeCalledTimes(1)
    expect(getVariation.mock.calls[0]).toEqual(['example', defaultEntity])
  })

  test('uses entity from prop "entity" if provided', () => {
    const getVariation = jest.fn(() => 'on')
    const defaultEntity = {id: '42'}
    const expectedEntity = {id: '9001'}
    const TestComponent = (props: any) => (
      <div>{JSON.stringify(props, null, 2)}</div>
    )
    const component = renderer.create(
      <flaggerCtx.Provider
        value={{
          entity: defaultEntity,
          loading: false,
          getVariation,
          getFlagDetails: jest.fn(),
        }}>
        <FlagSwitch
          flag="example"
          component={TestComponent}
          entity={expectedEntity}
        />
      </flaggerCtx.Provider>
    )

    expect(component.toJSON()).toMatchSnapshot()
    expect(getVariation).toBeCalledTimes(1)
    expect(getVariation.mock.calls[0]).toEqual(['example', expectedEntity])
  })

  test('filters children that has prop "case" equal to current flag variation', () => {
    const entity = {id: '9001'}
    const Child = (props: any) => <div>{JSON.stringify(props, null, 2)}</div>

    const component = renderer.create(
      <flaggerCtx.Provider
        value={{
          loading: false,
          getVariation: jest.fn(() => 'red'),
          getFlagDetails: jest.fn(),
        }}>
        <FlagSwitch flag="color" entity={entity}>
          <Child case="red" other="prop 1" />
          <Child case="red" other="prop 1.2" />
          <Child case="green" other="prop 2" />
          <Child case="blue" other="prop 3" />
        </FlagSwitch>
      </flaggerCtx.Provider>
    )
    const expectedRender = renderer.create(
      <>
        <Child
          case="red"
          other="prop 1"
          isSwitchChild={true}
          entity={entity}
          flag="color"
          variation="red"
          loading={false}
        />
        <Child
          case="red"
          other="prop 1.2"
          isSwitchChild={true}
          entity={entity}
          flag="color"
          variation="red"
          loading={false}
        />
      </>
    )

    expect(component.toJSON()).toMatchSnapshot()
    expect(component.toJSON()).toEqual(expectedRender.toJSON())
  })
})
