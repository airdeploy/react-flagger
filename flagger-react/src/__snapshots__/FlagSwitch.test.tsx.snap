// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FlagSwitch> tests accepts function as children 1`] = `
<div>
  {
  "flag": "example",
  "variation": "off",
  "loading": false
}
</div>
`;

exports[`<FlagSwitch> tests accepts react component as "component" prop 1`] = `
<div>
  {
  "flag": "example",
  "variation": "off",
  "loading": false
}
</div>
`;

exports[`<FlagSwitch> tests calls "getVariation()" from flaggerCtx 1`] = `
<div>
  {
  "flag": "example",
  "variation": "on",
  "loading": false
}
</div>
`;

exports[`<FlagSwitch> tests defaults to entity from flaggerCtx 1`] = `
<div>
  {
  "entity": {
    "id": "42"
  },
  "flag": "example",
  "variation": "on",
  "loading": false
}
</div>
`;

exports[`<FlagSwitch> tests filters children that has prop "case" equal to current flag variation 1`] = `
Array [
  <div>
    {
  "case": "red",
  "other": "prop 1",
  "isSwitchChild": true,
  "entity": {
    "id": "9001"
  },
  "flag": "color",
  "variation": "red",
  "loading": false
}
  </div>,
  <div>
    {
  "case": "red",
  "other": "prop 1.2",
  "isSwitchChild": true,
  "entity": {
    "id": "9001"
  },
  "flag": "color",
  "variation": "red",
  "loading": false
}
  </div>,
]
`;

exports[`<FlagSwitch> tests gets "loading" from flaggerCtx 1`] = `
<div>
  {
  "flag": "example",
  "variation": "off",
  "loading": true
}
</div>
`;

exports[`<FlagSwitch> tests uses entity from prop "entity" if provided 1`] = `
<div>
  {
  "entity": {
    "id": "9001"
  },
  "flag": "example",
  "variation": "on",
  "loading": false
}
</div>
`;
